### PR TITLE
support build rpm&deb

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve? 
related issue: https://github.com/pingcap/tidb-helper/issues/1

### What is changed and how it works?
By add Makefile and necessary configs，we can support build tidb/tikv/pd for various platforms, and generate rpm/deb package contains all components to set up a tidb cluster.

### Check List

Tests 

Code changes
- Makefile 
- configs(service/sysconfig/dockerfile)
- scripts for generate builder docker image

Note:
pr https://github.com/pingcap/tidb-helper/pull/2 add support for build for multi version os, please also take a look.

### CODE REVIEW

**Comment by glorv** [comment]
@Hoverbear PTAL

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+TIKV_SOURCE=$(BUILD_DIR)/tikv
+PD_SOURCE=$(BUILD_DIR)/pd
+ARTIFACT_BINARY=$(BUILD_DIR)/bin
+ARTIFACT_DIR=$(BUILD_DIR)/dist
+ARTIFACT_DOCKER=${ARTIFACT_DIR}/tidb-docker.tar.gz
+ARTIFACT_PACKAGE=$(ARTIFACT_DIR)/tidb-pkg
+ARTIFACT_RPM=${ARTIFACT_DIR}/
+DOCKER_IMAGE_NAME=pingcap/tidb:$(TAG)
+
+define fetch_source
```
Comment (Hoverbear) [inline Makefile:12]: Oh I didn't know you could do this! So cool!

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	else \
+		git clone https://github.com/pingcap/$(2).git $(1); \
+	fi
+endef
+
+$(BUILD_DIR):
+	mkdir -p $(BUILD_DIR)
+
+# don't use directory so we can force update the projects each time.
+TIDB_SOURCE:
```
Comment (Hoverbear) [inline Makefile:26]: Tasks without a target path should have `.PHONY: TIDB_SOURCE` which tells make that if a file called `./TIDB_SOURCE` exists it should still do this task.

File: etc/service/pd-server.service [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+Description=PD is the abbreviation for Placement Driver. It is used to manage and schedule the TiKV cluster.
+Wants=network-online.target
+
+[Service]
+User=tidb
+ExecStart=/usr/bin/pd-server --config $CONFIGURATION_DIRECTORY/pd/config.toml
+ExecReload=/bin/kill -HUP $MAINPID
+Restart=on-failure
+KillSignal=SIGINT
+LimitNOFILE=82920
```
Comment (Hoverbear) [inline etc/service/pd-server.service]: Seems @tennix suggests we tune this down.

File: etc/service/pd-server.service [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+ConfigurationDirectory=pd
+WorkingDirectory=/var/lib/pd/
+
+[Install]
+WantedBy=multi-user.target
```
Comment (Hoverbear) [inline etc/service/pd-server.service]: Missing newline

File: etc/service/tikv-server.service [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+ExecStart=/usr/bin/tikv-server --config $CONFIGURATION_DIRECTORY/tikv/config.toml
+ExecReload=/bin/kill -HUP $MAINPID
+Restart=on-failure
+KillSignal=SIGINT
+LimitNOFILE=82920
+ConfigurationDirectory=tikv
+WorkingDirectory=/var/lib/tikv/
+
+[Install]
+WantedBy=multi-user.target
```
Comment (Hoverbear) [inline etc/service/tikv-server.service]: Missing newline

File: etc/sysconfig/tidb-server.sysconfig [hunk 6]
```diff
@@ -0,0 +1 @@
+OPTIONS="-config /etc/tidb/config.toml"
```
Comment (Hoverbear) [inline etc/sysconfig/tidb-server.sysconfig]: This seems like an unnecessary layer of indirection

File: scripts/gen-rpm-spec.sh [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+%config %{_sysconfdir}/pd/config.toml.example
+%dir %{_sysconfdir}/pd
+%dir %attr(0755, tidb, tidb) %{_sharedstatedir}/pd
+%dir %attr(0755, tidb, tidb) %{_localstatedir}/log/pd
+
+%doc README.md
+%license LICENSE
+
+%changelog
+EOF
```
Comment (Hoverbear) [inline scripts/gen-rpm-spec.sh]: Missing newline

**Comment by Hoverbear** [comment]
I suggest the services be called `tidb` and `tikv` instead of `tidb-server` and `tikv-server` etc.

File: etc/service/tikv-server.service [hunk 8]
```diff
@@ -0,0 +1,16 @@
+[Unit]
```
Comment (Hoverbear) [inline etc/service/tikv-server.service:1]: You can copy this as part of the build process from the TiKV repo so if we update it you get the newer one. :)

File: Makefile [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+	rm -rf ${ARTIFACT_PACKAGE}
+
+.PHONY: clean-dist clean-bin clean-all
+clean-dist:
+	rm -rf build/dist
+
+clean-bin:
+	rm -rf build/bin
+
+clean-all:
```
Comment (Hoverbear) [inline Makefile]: Convention in Makefiles is `make clean` cleans. :)

**Comment by Hoverbear** [comment]
Great job, happy to give a LGTM once comments are resolved. :)

**Comment by glorv** [comment]
@Hoverbear  PTAL again. There is one issue left that the `tikv-server.service` should depend on the one in tikv project, but this can only be done after https://github.com/tikv/tikv/pull/6151 is merged.

File: etc/service/pd-server.service (hunk 3 above)
Comment (sykp241095) [reply etc/service/pd-server.service]: @Hoverbear 65535 ?

**Review by sykp241095 — APPROVED** [review]
LGTM

File: etc/service/pd-server.service (hunk 3 above)
Comment (Hoverbear) [reply etc/service/pd-server.service]: According to @tennix, only 1024. https://github.com/tikv/tikv/pull/6151#discussion_r357492785

**Comment by glorv** [comment]
@Hoverbear  It seems even if https://github.com/tikv/tikv/pull/6151 this pr is merged, if it doesn't cherry-pick to release-xx branch, we need to remain the `tikv-server.service` for a while so we can support build for tikv release branches.
I also add https://github.com/pingcap/tidb-helper/pull/3/files#diff-042fa8455879b862e2de2ee8130762e7R129-R130 these two lines to compress tikv-server/tikv-ctl as they don't exist in release branches yet.

**Pull request merged**